packages/flutter/lib/src/rendering/editable.dart

```diff
@@ -471,1767 +439,207 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin,
+  /// inside the text field viewport, as opposed to whether it is actually
+  /// visible on the screen.
+  ValueListenable<bool> get selectionEndInViewport => _selectionEndInViewport;
+  final ValueNotifier<bool> _selectionEndInViewport = ValueNotifier<bool>(true);
 
-    // When the text is obscured, the whole thing is treated as one big word.
-    if (obscureText) {
-      return moveSelectionToEnd(cause);
-    }
+  // Start TextMetrics.
 
-    assert(
-      _textLayoutLastMaxWidth == constraints.maxWidth &&
-      _textLayoutLastMinWidth == constraints.minWidth,
-      'Last width ($_textLayoutLastMinWidth, $_textLayoutLastMaxWidth) not the same as max width constraint (${constraints.minWidth}, ${constraints.maxWidth}).',
-    );
-    final TextSelection nextSelection = _moveGivenSelectionRightByWord(
-      _textPainter,
-      selection!,
-      includeWhitespace,
-    );
-    if (nextSelection == selection) {
-      return;
+  /// {@macro flutter.services.TextMetrics.getLineAtOffset}
+  @override
+  TextSelection getLineAtOffset(String text, TextPosition position) {
+    debugAssertLayoutUpToDate();
+    final TextRange line = _textPainter.getLineBoundary(position);
+    if (position.offset >= line.end)
+      return TextSelection.fromPosition(position);
+    // If text is obscured, the entire string should be treated as one line.
+    if (obscureText) {
+      return TextSelection(baseOffset: 0, extentOffset: text.length);
     }
-    _setSelection(nextSelection, cause);
+    return TextSelection(baseOffset: line.start, extentOffset: line.end);
   }
 
-  /// Move the current [selection] to the end of the field.
-  ///
-  /// {@macro flutter.rendering.RenderEditable.cause}
-  ///
-  /// See also:
-  ///
-  ///   * [moveSelectionToStart], which is the same but in the opposite
-  ///     direction.
-  void moveSelectionToEnd(SelectionChangedCause cause) {
-    assert(selection != null);
-
-    if (selection!.isCollapsed && selection!.extentOffset == _plainText.length) {
-      return;
-    }
-    final TextSelection nextSelection = TextSelection.collapsed(
-      offset: _plainText.length,
-    );
-    _setSelection(nextSelection, cause);
+  /// {@macro flutter.painting.TextPainter.getWordBoundary}
+  @override
+  TextRange getWordBoundary(TextPosition position) {
+    return _textPainter.getWordBoundary(position);
   }
 
-  /// Move the current [selection] to the start of the field.
-  ///
-  /// {@macro flutter.rendering.RenderEditable.cause}
-  ///
-  /// See also:
-  ///
-  ///   * [moveSelectionToEnd], which is the same but in the opposite direction.
-  void moveSelectionToStart(SelectionChangedCause cause) {
-    assert(selection != null);
+  /// {@macro flutter.services.TextMetrics.getTextPositionVertical}
+  @override
+  TextPosition getTextPositionVertical(int textOffset, double verticalOffset) {
+    final Offset caretOffset = _textPainter.getOffsetForCaret(TextPosition(offset: textOffset), _caretPrototype);
```

Comment:
This is where the one frame delay I mentioned in the doc. the _textPainter.text may be out of date from the text editing value. The caretOffset, or any other metrics render editable collected from the textpainter, may be out of date for one frame. I added a hack in _setSelection to get around this problem, but I am not sure whether there will be other corner case

Reply:
Ah I see, thanks for the explanation.  I'll take another look at what we can do to mitigate that problem after this is merged.